features/net/network-socket/WiFiInterface.h

```diff
@@ -43,28 +45,78 @@ class WiFiInterface: public NetworkInterface
 public:
     /** Start the interface
      *
-     *  Attempts to connect to a WiFi network. If passphrase is invalid,
-     *  NSAPI_ERROR_AUTH_ERROR is returned.
+     *  Attempts to connect to a WiFi network.
      *
      *  @param ssid      Name of the network to connect to
      *  @param pass      Security passphrase to connect to the network
      *  @param security  Type of encryption for connection
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+    virtual nsapi_error_t connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+
+    /** Start the interface
+     *
+     *  Attempts to connect to a WiFi network asynchronously, the call will return straight away. If the @a cb was NULL
+     *  you'll need to query @a get_state until it's in NSAPI_IF_STATE_CONNECTED state, otherwise the @a cb will be
+     *  called with connection results.
+     *
+     *  Note: @a ssid and @a pass must be kept until the connection is made, that is the callback has been called or the
+     *        state changed to @a NSAPI_IF_STATE_CONNECTED as they are passed by value.
+     *
+     *  @param ssid      Name of the network to connect to
+     *  @param pass      Security passphrase to connect to the network
+     *  @param security  Type of encryption for connection
+     *  @param cb        Function to be called when the connect finishes
+     *  @param data      Arbitrary user data to pass to @a cb function
+     */
+    virtual void connect_async(const char *ssid, const char *pass,nsapi_security_t security = NSAPI_SECURITY_NONE,
+                               wifi_connect_cb_t cb = NULL, void *data = NULL) = 0;
 
     /** Stop the interface
      *
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int disconnect() = 0;
+    virtual nsapi_error_t disconnect() = 0;
```

Comment:
Small request: Can we hold off on adopting the `nsapi_error_t` in this pr?

I agree it should be adopted for clarity, but the return types should be kept consistent untile then to avoid confusion.

IMO: We should make a separate pr to adopt these types across the entire network-socket API:
```
nsapi_error_t
nsapi_size_t
nsapi_size_or_error_t
```